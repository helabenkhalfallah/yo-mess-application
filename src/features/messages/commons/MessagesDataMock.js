const messagesFakeAction = {
  payload: {
    type: 'public',
  },
};

const messageDeleteFakeAction = {
  payload: {
    idMessage: '87654321',
  },
};

const messageAddFakeAction = {
  payload: {
    messageType: 'public',
    messageText: '23 ème édition du rallye Coeur de France, youpi !',
    messageSender: 'seancarr@hotmail.fr',
    messageReceiver: 'seancarr@hotmail.fr',
  },
};

const messagesFakeResponse = {
  data: [
    {
      idMessage: '87654321',
      messageType: 'public',
      messageDate: '08/10/2020',
      messageText: '23 ème édition du rallye Coeur de France, youpi !',
      messageSender: {
        userMail: 'seancarr@hotmail.fr',
        userFirstName: 'Sean',
        userLastName: 'Carr',
      },
      messageReceiver: null,
    }, {
      idMessage: '987654321',
      messageType: 'public',
      messageDate: '08/10/2020',
      messageText: 'Météo semaine prochaine : comme un avant-goût d\'hiver, le vent de Nord maintiendra les températures à un niveau inférieur de 5°C aux normales de saison, brrrrrr, préparez vos couettes :)))',
      messageSender: {
        userMail: 'isabellehoward@hotmail.fr',
        userFirstName: 'Isabelle',
        userLastName: 'Howard',
      },
      messageReceiver: null,
    },
  ],
};

const currentUser = {
  userToken: 'WwxxsQQSOPDHDG123g',
  userFirstName: 'Jean',
  userLastName: 'Dupont',
  userAddresse: '22 Avenue Victore Hugo, 94120, Fontenay Sous Bois',
  userPhone: '96222354145',
};

const userFriends = [
  {
    userMail: 'mellalaurent@hotmail.fr',
    userFirstName: 'Mella',
    userLastName: 'Laurent',
  },
  {
    userMail: 'isabellehoward@hotmail.fr',
    userFirstName: 'Isabelle',
    userLastName: 'Howard',
  },
  {
    userMail: 'seancarr@hotmail.fr',
    userFirstName: 'Sean',
    userLastName: 'Carr',
  },
];

const MessagesDataMock = {
  messagesFakeResponse,
  messageDeleteFakeAction,
  messagesFakeAction,
  messageAddFakeAction,
  currentUser,
  userFriends,
};

export default MessagesDataMock;
