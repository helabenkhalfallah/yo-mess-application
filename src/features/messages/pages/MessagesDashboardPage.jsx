import React, { Component, } from 'react';
import loadable from '@loadable/component';
import {
  Spin,
  Result,
} from 'antd';
import {
  ReduxMe,
  Matcher,
} from '../../../core';
import {
  AppLayout,
  withAuthentication,
} from '../../commons';
import {
  UserProvider,
  UserDispatcher,
} from '../../users/redux';
import {
  MessagesProvider,
  MessagesDispatcher,
} from '../redux';
import MessagesPropTypes from '../commons/MessagesPropTypes';

// dynamic load tabs content (load at need)
const MessagesList = loadable((props) => import(`../components/${props.path}`), {
  fallback: <Spin spinning />,
  cacheKey: (props) => props.path,
});

const MessageAddView = loadable((props) => import(`../components/${props.path}`), {
  fallback: <Spin spinning />,
  cacheKey: (props) => props.path,
});

const {
  UserFragment,
  UserLogoutFragment,
  UserFriendsFragment,
} = UserProvider;

const {
  MessagesFragment,
  MessageAddFragment,
  MessageDeleteFragment,
} = MessagesProvider;

@ReduxMe((state) => ({
  ...UserFragment(state),
  ...UserLogoutFragment(state),
  ...UserFriendsFragment(state),
  ...MessagesFragment(state),
  ...MessageAddFragment(state),
  ...MessageDeleteFragment(state),
}), {
  ...UserDispatcher,
  ...MessagesDispatcher,
})
@withAuthentication
class MessagesDashboardPage extends Component {
  currentType = 'public';

  constructor(props) {
    super(props);
    this.state = {
      showAddMessage: false,
    };
  }

  componentDidMount() {
    this.initPage();
  }

  initPage = () => {
    const {
      requestUserProfile,
      requestUserFriends,
    } = this.props;
      // default request public messages
    this.showPublicMessages();

    // request user profile if page is reloaded
    // and user still connected
    requestUserProfile();

    // prepare friends list before use
    requestUserFriends();
  }

  onMenuSelected = (menu) => {
    Matcher()
      .on(() => menu && menu === 'home', () => this.showPublicMessages())
      .on(() => menu && menu === 'messages', () => this.showPrivateMessages())
      .on(() => menu && menu === 'logout', () => this.onLogOut())
      .otherwise(() => {});
  }

  showPublicMessages = () => {
    this.currentType = 'public';
    const {
      requestMessages,
    } = this.props;
    requestMessages({
      type: 'public',
    });
  }

  showPrivateMessages = () => {
    this.currentType = 'private';
    const {
      requestMessages,
    } = this.props;
    requestMessages({
      type: 'private',
    });
  }

  onLogOut = () => {
    const {
      requestLogOutUser,
    } = this.props;
    requestLogOutUser();
  }

  openNewMessage = () => {
    this.setState({
      showAddMessage: true,
    });
  };

  hideNewMessage = () => {
    this.setState({
      showAddMessage: false,
    });
  };

  onSendMessage = (message) => {
    // eslint-disable-next-line no-console
    console.log('onSendMessage message : ', message);
    this.hideNewMessage();
    const {
      requestAddMessage,
    } = this.props;
    requestAddMessage(message);
  };

  render() {
    const {
      user,
      messages,
      messagesLoading,
      messagesError,
      userFriends,
    } = this.props;
    const {
      showAddMessage,
    } = this.state;
    return (
      <AppLayout
        selectedMenu="home"
        onNewMessage={this.openNewMessage}
        onMenuSelected={this.onMenuSelected}
        content={(
          <Spin
            spinning={messagesLoading}
            size="large"
          >
            {messagesError ? (
              <Result
                status="error"
                title="YoMess !"
                subTitle={messagesError}
              />
            ) : (
              <div className="user_form_login--page">
                <MessagesList
                  path="MessagesList"
                  messages={messages}
                />
                {showAddMessage && (
                  <MessageAddView
                    path="MessageAddView"
                    type={this.currentType}
                    showAddMessage={showAddMessage}
                    currentUser={user}
                    userFriends={userFriends}
                    onSendMessage={this.onSendMessage}
                    onCancel={this.hideNewMessage}
                  />
                )}
              </div>
            )}
          </Spin>
        )}
      />
    );
  }
}

MessagesDashboardPage.propTypes = {
  ...MessagesPropTypes.propTypes,
};

MessagesDashboardPage.defaultProps = {
  ...MessagesPropTypes.defaultProps,
};

export default MessagesDashboardPage;
