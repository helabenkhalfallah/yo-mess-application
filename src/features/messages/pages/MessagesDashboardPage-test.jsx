import React from 'react';
import { Provider, } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallow, } from 'enzyme';
import MessagesDataMock from '../commons/MessagesDataMock';
import MessagesDashboardPage from './MessagesDashboardPage';

const {
  messagesFakeResponse,
} = MessagesDataMock;

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('Test Suite - MessagesDashboardPage', () => {
  const initialState = {};
  it('Should render loading state', () => {
    const loadingMessages = {
      messages: {
        messages: null,
        loading: true,
        error: null,
      },
    };
    const store = mockStore({
      ...initialState,
      ...loadingMessages,
    });
    const wrapper = shallow(
      <Provider store={store}>
        <MessagesDashboardPage
          messagesLoading
          user={null}
          messages={null}
          messagesError={null}
          userFriends={null}
        />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('Should render messages', () => {
    const messagesState = {
      messages: {
        messages: messagesFakeResponse,
        loading: false,
        error: null,
      },
    };
    const store = mockStore({
      ...initialState,
      ...messagesState,
    });
    const wrapper = shallow(
      <Provider store={store}>
        <MessagesDashboardPage
          messagesLoading={false}
          user={null}
          messages={messagesFakeResponse}
          messagesError={null}
          userFriends={null}
        />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('Should render error', () => {
    const error = 'Une erreur technique c\'est produite, veuillez r√©essayez plus tard.';
    const messagesErrorState = {
      messages: {
        messages: null,
        loading: false,
        error,
      },
    };
    const store = mockStore({
      ...initialState,
      ...messagesErrorState,
    });
    const wrapper = shallow(
      <Provider store={store}>
        <MessagesDashboardPage
          messagesLoading={false}
          user={null}
          messages={null}
          messagesError={error}
          userFriends={null}
        />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
