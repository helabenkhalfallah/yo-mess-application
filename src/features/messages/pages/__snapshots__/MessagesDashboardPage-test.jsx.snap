// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Suite - MessagesDashboardPage Should render error 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(AuthenticationComponent)
    history={Object {}}
    messages={null}
    messagesError="Une erreur technique c'est produite, veuillez réessayez plus tard."
    messagesLoading={false}
    requestLogOutUser={null}
    requestLoginUser={null}
    requestMessages={null}
    requestUserFriends={null}
    requestUserProfile={null}
    user={null}
    userFriends={null}
    userLoginError={null}
    userLoginLoading={false}
  />
</ContextProvider>
`;

exports[`Test Suite - MessagesDashboardPage Should render loading state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(AuthenticationComponent)
    history={Object {}}
    messages={null}
    messagesError={null}
    messagesLoading={true}
    requestLogOutUser={null}
    requestLoginUser={null}
    requestMessages={null}
    requestUserFriends={null}
    requestUserProfile={null}
    user={null}
    userFriends={null}
    userLoginError={null}
    userLoginLoading={false}
  />
</ContextProvider>
`;

exports[`Test Suite - MessagesDashboardPage Should render messages 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(AuthenticationComponent)
    history={Object {}}
    messages={
      Object {
        "data": Array [
          Object {
            "idMessage": "87654321",
            "messageDate": "08/10/2020",
            "messageReceiver": null,
            "messageSender": Object {
              "userFirstName": "Sean",
              "userLastName": "Carr",
              "userMail": "seancarr@hotmail.fr",
            },
            "messageText": "23 ème édition du rallye Coeur de France, youpi !",
            "messageType": "public",
          },
          Object {
            "idMessage": "987654321",
            "messageDate": "08/10/2020",
            "messageReceiver": null,
            "messageSender": Object {
              "userFirstName": "Isabelle",
              "userLastName": "Howard",
              "userMail": "isabellehoward@hotmail.fr",
            },
            "messageText": "Des vélos électriques proposés en location longue durée !",
            "messageType": "public",
          },
        ],
      }
    }
    messagesError={null}
    messagesLoading={false}
    requestLogOutUser={null}
    requestLoginUser={null}
    requestMessages={null}
    requestUserFriends={null}
    requestUserProfile={null}
    user={null}
    userFriends={null}
    userLoginError={null}
    userLoginLoading={false}
  />
</ContextProvider>
`;
